File: finance-ui/backend/routers/schemas.py
"""
Schema management router
"""
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from database import get_db
from models.user import User
from models.schema import WorkType, SchemaStatus
from schemas.schema import (
    SchemaCreate, SchemaUpdate, SchemaResponse, SchemaDetailResponse,
    SchemaListResponse, TypeKeyRequest, TypeKeyResponse, NameExistsResponse,
    SchemaContentValidateRequest, SchemaContentValidateResponse,
    SchemaTestRequest, SchemaTestResponse
)
from services.schema_service import SchemaService
from routers.auth import get_current_user
from typing import Optional

router = APIRouter(prefix="/schemas", tags=["Schemas"])


@router.post("", response_model=SchemaResponse, status_code=201)
def create_schema(
    schema_data: SchemaCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Create a new schema
    """
    schema = SchemaService.create_schema(db, current_user, schema_data)
    return SchemaResponse.model_validate(schema)


@router.get("", response_model=SchemaListResponse)
def list_schemas(
    work_type: Optional[WorkType] = Query(None, description="Filter by work type"),
    status: Optional[SchemaStatus] = Query(None, description="Filter by status"),
    skip: int = Query(0, ge=0, description="Skip records"),
    limit: int = Query(100, ge=1, le=1000, description="Limit records"),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    List user's schemas
    """
    schemas, total = SchemaService.get_user_schemas(
        db, current_user, work_type, status, skip, limit
    )

    return SchemaListResponse(
        total=total,
        schemas=[SchemaResponse.model_validate(s) for s in schemas]
    )


@router.get("/{schema_id}", response_model=SchemaDetailResponse)
def get_schema(
    schema_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get schema details including content
    """
    schema = SchemaService.get_schema_by_id(db, current_user, schema_id)
    schema_content = SchemaService.get_schema_content(schema)

    response = SchemaDetailResponse.model_validate(schema)
    response.schema_content = schema_content

    return response


@router.put("/{schema_id}", response_model=SchemaResponse)
def update_schema(
    schema_id: int,
    update_data: SchemaUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Update schema
    """
    schema = SchemaService.update_schema(db, current_user, schema_id, update_data)
    return SchemaResponse.model_validate(schema)


@router.delete("/{schema_id}")
def delete_schema(
    schema_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Delete schema
    """
    SchemaService.delete_schema(db, current_user, schema_id)
    return {"message": "Schema deleted successfully"}


@router.post("/generate-type-key", response_model=TypeKeyResponse)
def generate_type_key(
    request: TypeKeyRequest,
    current_user: User = Depends(get_current_user)
):
    """
    Generate type_key from Chinese name using pinyin
    """
    type_key = SchemaService.generate_type_key_from_chinese(request.name_cn)
    return TypeKeyResponse(type_key=type_key)


@router.get("/check-name-exists", response_model=NameExistsResponse)
def check_name_exists(
    name_cn: str = Query(..., description="Chinese name to check"),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Check if name_cn already exists for the current user
    """
    exists = SchemaService.check_name_exists(db, current_user.id, name_cn)
    return NameExistsResponse(exists=exists)


@router.post("/validate-content", response_model=SchemaContentValidateResponse)
def validate_schema_content(
    request: SchemaContentValidateRequest,
    current_user: User = Depends(get_current_user)
):
    """
    Validate schema configuration structure and rules
    """
    result = SchemaService.validate_schema_content(request.schema_content)
    return result


@router.post("/test", response_model=SchemaTestResponse)
async def test_schema(
    request: SchemaTestRequest,
    current_user: User = Depends(get_current_user)
):
    """
    Test schema execution with uploaded files
    """
    result = await SchemaService.test_schema_execution(
        request.schema_content,
        request.file_paths
    )
    return result

File: finance-ui/src/api/schemas.ts
/**
 * Schema API
 */
import apiClient from './client';
import {
  Schema,
  SchemaDetail,
  CreateSchemaRequest,
  UpdateSchemaRequest,
  SchemaListResponse,
  SchemaFilters,
} from '@/types/schema';

export const schemaApi = {
  /**
   * Get list of schemas
   */
  getSchemas: async (filters?: SchemaFilters): Promise<SchemaListResponse> => {
    const response = await apiClient.get<SchemaListResponse>('/schemas', {
      params: filters,
    });
    return response.data;
  },

  /**
   * Create a new schema
   */
  createSchema: async (data: CreateSchemaRequest): Promise<Schema> => {
    const response = await apiClient.post<Schema>('/schemas', data);
    return response.data;
  },

  /**
   * Get schema details
   */
  getSchema: async (id: number): Promise<SchemaDetail> => {
    const response = await apiClient.get<SchemaDetail>(`/schemas/${id}`);
    return response.data;
  },

  /**
   * Update schema
   */
  updateSchema: async (id: number, data: UpdateSchemaRequest): Promise<Schema> => {
    const response = await apiClient.put<Schema>(`/schemas/${id}`, data);
    return response.data;
  },

  /**
   * Delete schema
   */
  deleteSchema: async (id: number): Promise<void> => {
    await apiClient.delete(`/schemas/${id}`);
  },

  /**
   * Generate type_key from Chinese name
   */
  generateTypeKey: async (nameCn: string): Promise<{ type_key: string }> => {
    const response = await apiClient.post<{ type_key: string }>('/schemas/generate-type-key', {
      name_cn: nameCn,
    });
    return response.data;
  },

  /**
   * Check if name_cn already exists for current user
   */
  checkNameExists: async (nameCn: string): Promise<{ exists: boolean }> => {
    const response = await apiClient.get<{ exists: boolean }>('/schemas/check-name-exists', {
      params: { name_cn: nameCn },
    });
    return response.data;
  },

  /**
   * Validate schema content structure
   */
  validateSchema: async (schemaContent: any): Promise<{
    valid: boolean;
    errors: string[];
    warnings: string[];
  }> => {
    const response = await apiClient.post('/schemas/validate-content', {
      schema_content: schemaContent,
    });
    return response.data;
  },

  /**
   * Test schema execution with files
   */
  testSchema: async (
    schemaContent: any,
    filePaths: string[]
  ): Promise<{
    success: boolean;
    output_preview: any[][];
    errors: string[];
    execution_time: number;
  }> => {
    const response = await apiClient.post('/schemas/test', {
      schema_content: schemaContent,
      file_paths: filePaths,
    });
    return response.data;
  },
};
